### ISSUE TITLE
TKMemo: RTF text pasting at very end of a kMemo

### ISSUE THREAD

**Issue by Freihut** [kryslt/KControls#69]
Using Laz 3.6 with fpc 3.2.2 and Kcontrols 1.7.3.0 ("up to date according to Laz' Online Package Manager") on Arch Linux. DE is XFCE, clipboard is handled by xfce4-clipman-plugin.

This happens if compiled for gtk2 or QT6 (others not yet tested). For some reason it depends on the RTF-text, but I didn't find the exact cause/trigger. 

But I was able to create a testcase:

1. Compile and start the example project [example.zip](https://github.com/user-attachments/files/17965003/example.zip)
2. Copy 1st line
3. Navigate to end of memo-box (via mouse or ctrl+end)
4. Paste via ctrl-v

(I tried this to automate via "Button 1" but it's barely tested.)

Expected result: TestImTheTextToPaste
Gotten result: ImTheTextToPasteTest

tl;dr: The appended text is pasted to the front of the line.

If you paste and copy this to an text-editor before continuing with Step 4 you'll get the expected result, so I guess this happens somewhere around 
```
TKCustomMemo.DoPaste()
[…]
LoadFromRTFStream(Stream, SelEnd);
```

and not in

`ActiveBlocks.InsertPlainText(SelEnd, S);`

I got this hotfixed by this patch, maybe this helps to find the cause: [uglybarelytestedhotfix.patch.zip](https://github.com/user-attachments/files/17965013/uglybarelytestedhotfix.patch.zip)

**Comment by kryslt** [comment]
Not a bug. 
Your RTF text is invalid. 
It misses a paragraph (a /par tag) at the end. 
Correct RTF input should be:

`KMemo1.RTF := '{\rtf1\ansi\ansicpg65001\deff0\uc1{\fonttbl{\f0\fcharset1\fprq0 Arial;}{\f1\fcharset0\fprq0 Droid Sans Armenian;}}{\colortbl\red255\green255\blue255;\red49\green55\blue57;\red0\green0\blue0;}{\*\listtable}{\*\listoverridetable}{\f1\fs20\cf1 ImTheTextToPaste }{\pard\ql\brdrcf2\f1\fs20\cf1\par}{\f1\fs20\cf1 Test}{\pard\ql\brdrcf2\f1\fs20\cf1\par}}';
`

What software did you use to produce such RTF?

Anyway, you may also fix this by adding paragraph manually:

`KMemo1.Blocks.AddParagraph();`

**Comment by Freihut** [comment]
Thanks for the reply, it helps a lot.

> What software did you use to produce such RTF?

Such invalid RTF came from one of my projects called "Buuugs", but it doesn't produce any RTF text it only passes it from a Kmemo to a database and vice versa.

**Comment by Freihut** [comment]
I did some investigations and this is how this invalid RTF text may have been created:

1. Comment out KMemo1.RTF := … from FormCreate() in the example or create an empty form with an TKMemo and a button which runs `ShowMessage(KMemo1.RTF);`
2. Compile and build
KMemo1.RTF at this time is
`{\rtf1\ansi\ansicpg65001\deff0\uc1{\fonttbl{\f0\fcharset1\fprq0 Arial;}}{\colortbl\red255\green255\blue255;\red26\green26\blue26;\red0\green0\blue0;}{\*\listtable}{\*\listoverridetable}{\pard\ql\brdrcf2\f0\fs20\cf1\par}}`
3. Write HenloTXT and hit enter to create a line break
KMemo1.RTF at this time is
`{\rtf1\ansi\ansicpg65001\deff0\uc1{\fonttbl{\f0\fcharset1\fprq0 Arial;}}{\colortbl\red255\green255\blue255;\red26\green26\blue26;\red0\green0\blue0;}{\*\listtable}{\*\listoverridetable}{\f0\fs20\cf1 HenloTXT}{\pard\ql\brdrcf2\f0\fs20\cf1\par}{\pard\ql\brdrcf2\f0\fs20\cf1\par}}`
4. Mark and copy Henlo via keyboard
5. Select everything behind TXT like this: 
![Screenshot_2024-18-04_12-14-23](https://github.com/user-attachments/assets/52c29933-5b68-4616-bf88-a67dabcdc9f2)
(orange = selection color)
6. Paste via keyboard
7. Text is now HenloHenloTXT and RTF is
`{\rtf1\ansi\ansicpg65001\deff0\uc1{\fonttbl{\f0\fcharset1\fprq0 Arial;}}{\colortbl\red255\green255\blue255;\red26\green26\blue26;}{\*\listtable}{\*\listoverridetable}{\f0\fs20\cf1 Henlo}{\f0\fs20\cf1 HenloTXT}}`

**Comment by kryslt** [comment]
Yes this is a bug indeed.
It is possible to delete the last paragraph by selecting it and using a clipboard operation (Cut or Paste) or with Delete/Backspace keys.
Text processors like Word never allow this. 
This must be fixed.

EDIT: Similar issue appears in a text box or a table cell, i.e. in each TKMemoContainer and its blocks.
